### PR TITLE
package: add generic FatFs

### PR DESCRIPTION
**Pull request by MichelRottleuthner**
This PR adds a package for a [generic FAT file system](http://elm-chan.org/fsw/ff/00index_e.html). As it's a RIOT package it mainly contains a makefile that can be used to fetch the code from the original authors page. 
Beyond this the PR also includes two FatFs abstraction layer implementations. One for the [sdcard_spi](https://github.com/RIOT-OS/RIOT/pull/6031) driver and one that wraps around file-io systemcalls that makes it fairly easy to use it with FAT-images under native (see [README](https://github.com/MichelRottleuthner/RIOT/blob/fatfs_testing/tests/pkg_fatfs/README.md) for details). 
Also included is a simple interactive test application. Currently with the test app you can mount your volume, list, read, create and write to files.
One thing I'd like to discuss about this thing: Since the FatFs package is generally neither bound to a specific media (like sdcard in this case) nor is the sdcard bound to be only used with FatFs - what would be the best and "RIOT-like" way to provide the abstraction layer (diskio.c)?. For now I've included it to this PR to make testing easier but I think it should be moved somewhere else.

Waiting for [6031](https://github.com/RIOT-OS/RIOT/pull/6031) so that building with sdcard_spi_diskio is possible.

### CODE REVIEW

**Comment by OlegHahm** [comment]
There was FAT support in former RIOT versions, but we decided to kick it out at some point. IIRC the main reasons were:
1.) Unclear license situation. It is not clear if all code in the ChaN implementation is correctly licensed.
2.) Not designed for simple flash storage without internal wear leveling.
3.) Horrible design of FAT FS itself.

Have you considered to use another FS that is more suited for MCUs and flash storage? Maybe you can take a look at the on-going discussions with respect to file systems: https://github.com/RIOT-OS/RIOT/search?q=fs&type=Issues

**Comment by MichelRottleuthner** [comment]
>1.) Unclear license situation.

[Hmm](http://elm-chan.org/fsw/ff/en/appnote.html#license) can you give me a pointer to where that was discussed? The only related info I could find about that in the RIOT context is within #2474 where @x3ro stats "custom license compatible to BSD"

>2.) Not designed for simple flash storage without internal wear leveling.

True, but I'd agree to comments by @miri64 like [this one](https://github.com/RIOT-OS/RIOT/pull/5616#issuecomment-232278749) or by @x3ro: "FAT would be useful for SD card storage, which provide internal wear levelling, but it would probably wear out simpler flash chips too quickly to be useful on those" [(found here)](https://github.com/RIOT-OS/RIOT/pull/2474). 
Apart from that, FatFs could be still a usefull addition especially for simple read-only tasks (config files with the ability to modify on any system and simply put back the sd-card to the MCU). Also if you just want to log data as in "continuously write data to a huge sd-card without deleting files for a long time" (my use case for now) wear-leveling might be not a real problem to you.

>3.) Horrible design of FAT FS itself.

Yep! Nevertheless some guys ( @alignan @smlng I'm pointing at you :P) were eager to try it out.

**Comment by OlegHahm** [comment]
There's some part of the discussion here: https://github.com/RIOT-OS/RIOT/pull/1265#issuecomment-71436272

**Comment by OlegHahm** [comment]
> lso if you just want to log data as in "continuously write data to a huge sd-card without deleting files for a long time" (my use case for now) wear-leveling might be not a real problem to you.

True, but in this case I don't see the need for any file system.

However, don't get me wrong: I'm not in general opposed to this PR.

Personally, I'm still not convinced that a file system is really needed on these systems, but that's another discussion. But if we want to support file systems, we should have a clear understanding of what and how we want to integrate it into RIOT.

**Comment by OlegHahm** [comment]
The only license issue I could find was related to long file names and a MS patent, but that doesn't seem to be a problem: http://elm-chan.org/fsw/ff/en/appnote.html#lfn

Maybe I remember wrong, but I will still try to see if anyone else remembers.

**Comment by alignan** [comment]
> Personally, I'm still not convinced that a file system is really needed on these systems

The point of using a micro-SD instead of a flash memory is to allow the information to be retrieved and read externally, without having to use a specific test/feature such as dumping the memory content over USB, etc (which would be huge).  Also, in cases external configuration information has to be written to the microSD cards, it would be easier just to clone a card.

I'm not particularly fixed to FatFS as the solution, but I do see a need for a file system for the Micro-SD

**Comment by OlegHahm** [comment]
> The point of using a micro-SD instead of a flash memory is to allow the information to be retrieved and read externally, without having to use a specific test/feature such as dumping the memory content over USB, etc (which would be huge). Also, in cases external configuration information has to be written to the microSD cards, it would be easier just to clone a card.

I see use cases for a memory card, but for simple configuration data or log information, one doesn't need a filesystem. Cloning a card would be done with something like `dd` anyway which doesn't care about a filesystem or not.

But let's not discuss this here. :wink:

**Comment by miri64** [comment]
> Cloning a card would be done with something like dd anyway which doesn't care about a filesystem or not.

What a user friendly way to get a bitmap to your device ;-)

**Comment by MichelRottleuthner** [comment]
>True, but in this case I don't see the need for any file system.

Sure. I simply did it for ease of use and the fact that i can manipulate files if needed - with the implication that I may need to think a bit more about how to use the FS.

>Personally, I'm still not convinced that a file system is really needed on these systems, but that's another discussion.

I agree. Its not "really" needed as in "one could always implement/port a specific fs if needed". But also it could be of great help for someone that just wants to store/load data in an interoperable way and wants to get started quickly without reinventing the weel on how to store the data on the bare media.

>...we should have a clear understanding of what and how we want to integrate it into RIOT

Agree.  And I'm totally open minded to any kind of suggestions to tackle this :)

>I'm not particularly fixed to FatFS as the solution, but I do see a need for a file system for the Micro-SD 

ACK! Would also like to see a better fs without all the shortcomings. For me the key thing is: FatFs is actually usable in it's current state and its dead simple (from the user POV). FAT has it's place and may still be there for longer than we'd all prefer so why not provide compatibility...

**Comment by kaspar030** [comment]
> Nope, wrong pointer. It might be that this discussion was in pre-RIOT, pre-Github days. I will see if I can find something. @kaspar030, do you remember?

AFAIR the guy doing FireKernel's FAT support didn't use chan or (had to) heavily modify it, so we kicked it out as unmaintainable.

I'd like to see FAT for RIOT, just because it's the common denominator all OS can read. Wear leveling shouldn't be a problem with SD cards as they deal with that internally.

**Comment by OlegHahm** [comment]
> AFAIR the guy doing FireKernel's FAT support didn't use chan or (had to) heavily modify it, so we kicked it out as unmaintainable.

I just checked: the latest version of FireKernel used ChaN FatFs R0.08a. I haven't run a diff versus the upstream version.

> I'd like to see FAT for RIOT, just because it's the common denominator all OS can read. 

It is so sad that in 2016 we still have to use FAT as common FS, but yeah, you're kind of right. However, it's interesting how drastically you've changed your mind: https://github.com/RIOT-OS/RIOT/pull/1265#issuecomment-71465603 :wink:

**Comment by kaspar030** [comment]
> It is so sad that in 2016 we still have to use FAT as common FS, but yeah, you're kind of right. However, it's interesting how drastically you've changed your mind: #1265 (comment)

Yup. Back then I remembered the horrors Michael went through to get FAT working. Now it seems as ChaN's FatFS is actually usable. ;)

**Comment by kaspar030** [comment]
@MichelRottleuthner Have you considered basing this on #5616?

**Comment by MichelRottleuthner** [comment]
Yes I think that would make sence but I haven't looked into the details so I'm not exactly sure what is needed to make this work with vfs.

**Comment by MichelRottleuthner** [comment]
@cgundogan @smlng you can now also use FatFs on native see README.md for details.

**Comment by cgundogan** [comment]
@MichelRottleuthner :+1: I will look into the native tests later today. Do you mind splitting the `sdcard_spi` based code parts into another PR, which depends on this PR? This would make it easier to review the FatFS+native relevant code.

**Comment by MichelRottleuthner** [comment]
> splitting the `sdcard_spi` based code parts into another PR

As expressed in the opening comment I'd like to do that anyway. The question that remains open for discussion is: how (i.e. where would be the best place to put this code)? 
If done consequently the gluecode for native would also have to be kicked out of this PR, which wouldn't really help with testing for now I think. 
Currently I think an additional module (fatfs-diskio or whatever) that is also included in this PR would be best to ease testing and keep it clean at the same time. What do the others think?

File: sys/fatfs_diskio/native/native_diskio.c [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        .fd = NULL,
+        .opened  = false
+    },
+};
+
+static inline dummy_volume_t *get_volume_file(uint32_t idx)
+{
+    if(idx < sizeof(volume_files) / sizeof(dummy_volume_t)){
+        return &volume_files[idx];
+    }else{
```
Comment (cgundogan) [inline sys/fatfs_diskio/native/native_diskio.c]: `else` statements should go to a new line

File: sys/fatfs_diskio/native/native_diskio.c [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+typedef struct {
+    const char *image_path;
+    FILE *fd;
+    bool opened;
+} dummy_volume_t;
+
+static dummy_volume_t volume_files[] = {
+    {
+        .image_path = "../../riot_fatfs_disk.img",
```
Comment (cgundogan) [inline sys/fatfs_diskio/native/native_diskio.c]: this should be parameterized somehow

File: sys/fatfs_diskio/native/native_diskio.c [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+ * @param[in]  count   Number of sectors to write
+ *
+ * @return             RES_OK if no error occurred
+ * @return             RES_NOTRDY if data wasn't written completely
+ */
+DRESULT disk_write (BYTE pdrv, const BYTE *buff, DWORD sector, UINT count)
+{
+    dummy_volume_t *volume = get_volume_file(pdrv);
+
+    if(volume != NULL && volume->opened){
```
Comment (cgundogan) [inline sys/fatfs_diskio/native/native_diskio.c]: put parens around separate conditions

File: sys/fatfs_diskio/native/native_diskio.c [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+#define GET_SECTOR_SIZE     2   
+
+/* Get erase block size (needed at _USE_MKFS == 1) */
+#define GET_BLOCK_SIZE      3   
+
+/* Inform device that the data on the block of sectors is 
+   no longer used (needed at _USE_TRIM == 1) */
+#define CTRL_TRIM           4   
+
+#define RTC_YEAR_OFFSET   1900
```
Comment (cgundogan) [inline sys/fatfs_diskio/native/native_diskio.c]: many of these macros are defined multiple times. can you extract them into a common h file?

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+static inline sd_card_t *get_sd_card(int idx){
+    if(idx < NUM_OF_SD_CARDS){
+        return &(cards[idx]);
+    }else{
+        return NULL;
+    }
+}
+
+#ifdef FATFS_RTC_AVAILABLE
+DWORD get_fattime (void){
```
Comment (cgundogan) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: seems to be the same function as above. would make sense to put all common functions / macros into a common file

Comment (cgundogan) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: https://github.com/RIOT-OS/RIOT/pull/6072/files#diff-e02348602796bab099bf4cd798e87f8eR81

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+/**
+ * @brief           returns the status of the disk
+ *
+ * @param[in] pdrv  drive number to identify the drive
+ *
+ * @return          STA_NODISK if no disk exists with the given id
+ * @return          0 if disk is initialized
+ * @return          STA_NOINIT if disk id exists, but disk isn't initialized 
+ */
+DSTATUS disk_status (BYTE pdrv)
```
Comment (cgundogan) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: same?

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+ * @param[in]  buff    Data to be written
+ * @param[in]  sector  Start sector in LBA
+ * @param[in]  count   Number of sectors to write
+ *
+ * @return             RES_OK if no error occurred
+ * @return             RES_NOTRDY if data wasn't written completely
+ */
+DRESULT disk_write (BYTE pdrv, const BYTE *buff, DWORD sector, UINT count)
+{
+    //printf("diskio.c: disk_write called!\n");
```
Comment (BytesGalore) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: please remove

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+                return RES_OK;
+            }else{
+                return RES_ERROR;   
+            }
+            
+        #endif
+
+        #if(_MAX_SS != _MIN_SS)
+        case GET_SECTOR_SIZE;
+            *buff = 512;
```
Comment (BytesGalore) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: please provide a define for this value

File: tests/pkg_fatfs/main.c [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+#define RTC_YEAR_OFFSET   1900
+#define FATFS_YEAR_OFFSET 1980
+#define TEST_FATFS_READ_BUFFER_SIZE 64
+#define TEST_FATFS_MAX_VOL_STR_LEN 8
+
+FATFS fat_fs;        /* FatFs work area needed for each volume */
+
+static int _mount(int argc, char **argv)
+{
+    printf("mounting file system image...");
```
Comment (BytesGalore) [inline tests/pkg_fatfs/main.c]: use puts please

File: tests/pkg_fatfs/main.c [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+    if (argc != 2) {
+        printf("usage: %s <filename>\n", argv[0]);
+        return -1;
+    }
+
+    if (f_open(&fd, argv[1], FA_WRITE | FA_CREATE_ALWAYS) == FR_OK) {
+        FRESULT close_resu = f_close(&fd);
+        if (close_resu == FR_OK) {
+            printf("[OK]\n");
```
Comment (BytesGalore) [inline tests/pkg_fatfs/main.c]: puts please

File: tests/pkg_fatfs/main.c [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+                printf("[FAILED] (f_read error %d)\n", read_resu);
+                resu = -4;
+                break;
+            }
+
+            for (uint32_t i = 0; i < read_chunk; i++) {
+                printf("%c", buffer[i]);
+            }
+        }
+        printf("\n");
```
Comment (BytesGalore) [inline tests/pkg_fatfs/main.c]: puts also here

File: sys/fatfs_diskio/native/native_diskio.c [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+#include <stdlib.h>
+#include <errno.h>
+
+/* Complete pending write process (needed at _FS_READONLY == 0) */
+#define CTRL_SYNC           0   
+
+/* Get media size (needed at _USE_MKFS == 1) */
+#define GET_SECTOR_COUNT    1   
+
+/* Get sector size (needed at ) */
```
Comment (BytesGalore) [inline sys/fatfs_diskio/native/native_diskio.c]: where?

File: sys/fatfs_diskio/native/native_diskio.c [hunk 13]
```diff
@@ -0,0 +1,287 @@
+/*
```
Comment (miri64) [inline sys/fatfs_diskio/native/native_diskio.c]: Maybe move this file to `cpu/native` instead, since it is only relevant for native anyway

Comment (miri64) [inline sys/fatfs_diskio/native/native_diskio.c]: Sorry, did not see the SPI implementation below.

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+ * @param[in] pdrv  drive number to identify the drive
+ *
+ * @return          STA_NODISK if no disk exists with the given id
+ * @return          0 if disk was initialized sucessfully
+ * @return          STA_NOINIT if disk id exists, but couldn't be initialized 
+ */
+DSTATUS disk_initialize (BYTE pdrv)
+{
+
+    cards[0].spi_dev = TEST_SDCARD_SPI;
```
Comment (miri64) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: This should be done, every time a disk is initialized?

File: sys/fatfs_diskio/native/native_diskio.c [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+ * @param[in]  count   Number of sectors to read
+ *
+ * @return             RES_OK if no error occurred
+ * @return             RES_NOTRDY if data wasn't read completely
+ */
+DRESULT disk_read (BYTE pdrv, BYTE *buff, DWORD sector, UINT count)
+{
+    dummy_volume_t *volume = get_volume_file(pdrv);
+
+    if(volume != NULL && volume->opened){
```
Comment (BytesGalore) [inline sys/fatfs_diskio/native/native_diskio.c]: please add brackets to separate the statements

File: sys/fatfs_diskio/native/native_diskio.c [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+ * @return                 RES_PARERR if an error occurred
+ */
+DRESULT disk_ioctl (
+    BYTE pdrv,      /*  */
+    BYTE cmd,       /*  */
+    void *buff      /* Buffer to send/receive control data */
+)
+{
+    (void) pdrv; /* prevent warning about unused param */
+    (void) buff; /* prevent warning about unused param */
```
Comment (BytesGalore) [inline sys/fatfs_diskio/native/native_diskio.c]: why unused?

File: tests/pkg_fatfs/Makefile [hunk 17]
```diff
@@ -0,0 +1,18 @@
+APPLICATION = pkg_fatfs
+include ../Makefile.tests_common
+
+USEMODULE += shell
+
```
Comment (smlng) [inline tests/pkg_fatfs/Makefile]: Please add `BOARD ?= native` directive

File: tests/pkg_fatfs/main.c [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+    }
+    return 0;
+}
+
+static int _read(int argc, char **argv)
+{
+    FIL fd;
+    int resu = 0;
+
+    if (argc != 2) {
```
Comment (smlng) [inline tests/pkg_fatfs/main.c]: fails if `len` parameter, because then `argc=3`

File: tests/pkg_fatfs/main.c [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+    }
+    else {
+        printf("[FAILED] (f_open error %d)\n", open_resu);
+        resu = -2;
+    }
+
+    return resu;
+}
+
+static int _write(int argc, char **argv)
```
Comment (smlng) [inline tests/pkg_fatfs/main.c]: does not append as suggested by `help`, but rather overwrites existing file content

**Comment by smlng** [comment]
@MichelRottleuthner very nice. Works on Debian 8 right out of the box, please add command reference for RIOT shell, i.e., only `mount 0` works. Tried to run this on macOS but failed already while patching, I'll look into it and report ...

File: pkg/fatfs/patches/disable_rtc.patch [hunk 20]
```diff
@@ -0,0 +1,4 @@
+218c218
+< #define _FS_NORTC	0
+---
+> #define _FS_NORTC	1
```
Comment (smlng) [inline pkg/fatfs/patches/disable_rtc.patch]: This patch fails on macOS, please change patch format/syntax as follows:

```
--- ffconf.h	2016-09-04 21:41:04.000000000 +0200
+++ ffconf_patched.h	2016-11-29 22:04:09.000000000 +0100
@@ -215,7 +215,7 @@
 /  Note that enabling exFAT discards C89 compatibility. */
 
 
-#define _FS_NORTC	0
+#define _FS_NORTC	1
 #define _NORTC_MON	1
 #define _NORTC_MDAY	1
 #define _NORTC_YEAR	2016
```

File: pkg/fatfs/Makefile.include [hunk 21]
```diff
@@ -0,0 +1,2 @@
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)
+DIRS += $(BINDIRBASE)/pkg/$(BOARD)/fatfs
```
Comment (smlng) [inline pkg/fatfs/Makefile.include]: Compile of fatfs fails on macOS with:
```
/RIOT/tests/pkg_fatfs/bin/pkg/native/fatfs/ff.c:4942:51: error: for loop has empty body
      [-Werror,-Wempty-body]
        for (slen = 0; (UINT)label[slen] >= ' '; slen++) ;      /* Get name length */
                                                         ^
/RIOT/tests/pkg_fatfs/bin/pkg/native/fatfs/ff.c:4942:51: note: put the semicolon on a
      separate line to silence this warning
```

please add the following to compile on macOS:
```
ifeq ($(shell uname -s),Darwin)
    CFLAGS += -Wno-empty-body
endif
```

File: tests/pkg_fatfs/main.c [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+
+            uint32_t fr_gib_i = free_bytes / (1024 * 1024 * 1024);
+            uint32_t fr_gib_f = ((((free_bytes/(1024 * 1024)) - fr_gib_i * 1024) * 1000) / 1024);
+
+            printf("%"PRIu32",%03"PRIu32" GiB of %"PRIu32",%03"PRIu32" \
+                    GiB available\n", fr_gib_i, fr_gib_f, to_gib_i, to_gib_f);
+        }
+    }
+    else {
+        printf("[FAILED]\n");
```
Comment (cgundogan) [inline tests/pkg_fatfs/main.c]: puts

Comment (cgundogan) [inline tests/pkg_fatfs/main.c]: here and following also

File: pkg/fatfs/Makefile.include (hunk 21 above)
Comment (MichelRottleuthner) [inline pkg/fatfs/Makefile.include]: would Makfile.include be the right place for that?

Comment (smlng) [inline pkg/fatfs/Makefile.include]: where else? The problem is the CFLAG must be added in any RIOT project that uses `fatfs`, adding these lines to the primary `Makefile` does not help (I tested btw.) because that one only downloads and patches the sources. But `Makefile.include` is recursivly added during build of any RIOT project, that's why it has to be in there.

File: sys/fatfs_diskio/native/Makefile [hunk 23]
```diff
@@ -0,0 +1,9 @@
+MODULE = fatfs_diskio_native
+
+INCLUDES += -I$(RIOTBASE)/sys/include/fatfs_diskio
+
+FATFS_DISKIO_NATIVE_DEFAULT_FILE ?= \"../../riot_fatfs_disk.img\"
```
Comment (smlng) [inline sys/fatfs_diskio/native/Makefile]: This should also be added to `tests/pkg_fatfs/Makefile`, further I think its not wise to look for a file 2 dirs up by default. This clearly depends where you build and exec the binary. Hence, should be just `riot_fatfs_disk.img`.

File: sys/fatfs_diskio/native/native_diskio.c [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+
+    if(volume == NULL){
+        return STA_NODISK;
+    }
+    if(volume->opened){
+        return 0;
+    }
+    else{
+        return STA_NOINIT;  
+    }
```
Comment (smlng) [inline sys/fatfs_diskio/native/native_diskio.c]: Coding style: spaces before braces. Here and everywhere 😄

File: sys/fatfs_diskio/native/native_diskio.c [hunk 25]
```diff
@@ -0,0 +1,274 @@
...
+
+    if((volume != NULL) && volume->opened){
+        /* set read pointer to secor equivalent position */
+        if(fseek(volume->fd, sector * FIXED_BLOCK_SIZE, SEEK_SET) == 0){
+            if(fread(buff, FIXED_BLOCK_SIZE, count, volume->fd) == count){
+                return RES_OK;
+            }
+            else{
+                DEBUG("diskio_native.c: disk_read: fread: \
+errno: 0x%08x\n", errno);
```
Comment (smlng) [inline sys/fatfs_diskio/native/native_diskio.c]: is this line break necessary? If so then align with parentheses of `DEBUG(...`

File: sys/fatfs_diskio/native/native_diskio.c [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+                    return RES_OK;
+                }
+                else{
+                    DEBUG("diskio_native.c: disk_write: fflush: \
+errno: 0x%08x\n", errno);    
+                }
+            }
+            else{
+                DEBUG("diskio_native.c: disk_write: fwrite: \
+errno: 0x%08x\n", errno);
```
Comment (smlng) [inline sys/fatfs_diskio/native/native_diskio.c]: Same for the debug message, see above

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+#include "debug.h"
+
+extern sd_card_t cards[NUM_OF_SD_CARDS];
+
+static inline sd_card_t *get_sd_card(int idx)
+{
+    if (idx < NUM_OF_SD_CARDS) {
+        return &(cards[idx]);
+    }
+    else {
```
Comment (smlng) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: simplify to `return NULL;`

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 28]
```diff
@@ -0,0 +1,204 @@
...
+ * @return          STA_NOINIT if disk id exists, but disk isn't initialized
+ */
+DSTATUS disk_status(BYTE pdrv)
+{
+    sd_card_t *card = get_sd_card(pdrv);
+
+    if (card == NULL) {
+        return STA_NODISK;
+    }
+    if (card->init_done) {
```
Comment (smlng) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: should be `else if`  as below

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 29]
```diff
@@ -0,0 +1,204 @@
...
+        return 0;
+    }
+    else {
```
Comment (smlng) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: simplify to `return STA_NOINIT;` without `else`

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 30]
```diff
@@ -0,0 +1,204 @@
...
+{
+    sd_card_t *card = get_sd_card(pdrv);
+
+    if (card == NULL) {
+        return STA_NODISK;
+    }
+    else if (sdcard_spi_init(card)) {
+        return 0;
+    }
+    else {
```
Comment (smlng) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: simplify to `return STA_NOINIT;` without `else`

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 31]
```diff
@@ -0,0 +1,204 @@
...
+ * @return             RES_OK if no error occurred
+ * @return             RES_NOTRDY if data wasn't read completely
+ */
+DRESULT disk_read(BYTE pdrv, BYTE *buff, DWORD sector, UINT count)
+{
+    sd_card_t *card = get_sd_card(pdrv);
+
+    if ((card != NULL) && card->init_done) {
+        sd_rw_response_t state;
+        int read = sdcard_spi_read_blocks(card, sector, (char *)buff,
```
Comment (smlng) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: rewrite to 
```
if () {
    sd_rw_response_t state;
    if((count != sdcard_spi_read_blocks(card, sector, (char *)buff, SD_HC_BLOCK_SIZE, count, &state))) {
        printf("disk_read: sdcard_spi_read_blocks: ERROR:%d\n", state);
        return RES_NOTRDY;
    }
}
return RES_OK;

File: sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c [hunk 32]
```diff
@@ -0,0 +1,204 @@
...
+ * @return             RES_NOTRDY if data wasn't written completely
+ */
+DRESULT disk_write(BYTE pdrv, const BYTE *buff, DWORD sector, UINT count)
+{
+    sd_card_t *card = get_sd_card(pdrv);
+
+    if ((card != NULL) && card->init_done) {
+        sd_rw_response_t state;
+        int written = sdcard_spi_write_blocks(card, sector, (char *)buff,
+                                              SD_HC_BLOCK_SIZE, count, &state);
```
Comment (smlng) [inline sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c]: could be rewritten, see above

File: sys/include/fatfs_diskio/fatfs_diskio_common.h [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+#define FATFS_DISKIO_COMMON_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "fatfs/diskio.h" /* FatFs lower layer API */
+
+/* Complete pending write process (needed when _FS_READONLY == 0) */
+#define CTRL_SYNC           0
```
Comment (smlng) [inline sys/include/fatfs_diskio/fatfs_diskio_common.h]: comment style: 
```
#define CTRL_SYNC           0 /*< Complete pending write process */
```

same below

**Comment by smlng** [comment]
@BytesGalore can you have another look on this one? I think its fine now and ready for merging, so I give murdock a try.

**Comment by BytesGalore** [comment]
@smlng I think it still needs an uncrustification and some minor beautification, e.g. `printf` -> `puts` and magic numbers -> defines

File: tests/pkg_fatfs/main.c [hunk 34]
```diff
@@ -126,15 +141,13 @@ static int _mk(int argc, char **argv)
...
-    }
-    else {
-        printf("failed to create file '%s'\n", argv[1]);
+
+        printf("[FAILED] (f_close error %d)\n", close_resu);
+        return -2;
     }
-    return 0;
+
+    printf("failed to create file '%s'\n", argv[1]);
```
Comment (smlng) [inline tests/pkg_fatfs/main.c]: minor nit here: all other error messages start with `[FAILED] `, I suggest to adapt wording accordingly?

**Comment by PeterKietzmann** [comment]
`/data/riotbuild/sys/fatfs_diskio/sdcard_spi/sdcard_spi_diskio.c:25:24: fatal error: sdcard_spi.h: No such file or directory #include "sdcard_spi.h"`

**Comment by MichelRottleuthner** [comment]
@PeterKietzmann that's the reason for "waiting for #6031" ;)

**Comment by smlng** [comment]
@PeterKietzmann I think @MichelRottleuthner should whitelist supported boards with SPI as defined in [pkg test Makefile](https://github.com/MichelRottleuthner/RIOT/blob/5c9f6707e56c73c1f2ed6403037bb7d07a9b4880/tests/pkg_fatfs/Makefile)

**Comment by smlng** [comment]
plus dependence on #6031

**Comment by miri64** [comment]
@MichelRottleuthner please rebase.

**Comment by miri64** [comment]
VFS (#5616) was merged, so if you want to you can add a wrapper for that, but given that this PR was already ACK'd, I would say we do this as a follow-up ;-)

**Comment by miri64** [comment]
Was approved, but needs rebase and some fixing ;-)

File: sys/fatfs_diskio/Makefile [hunk 35]
```diff
@@ -0,0 +1,2 @@
+MODULE = fatfs_diskio
+include $(RIOTBASE)/Makefile.base
```
Comment (miri64) [inline sys/fatfs_diskio/Makefile]: If this module is fatfs specific: why are they in `sys` and not `pkg/fatfs/`? Please move them (sorry for ringing in so late, but just noticed).

Comment (MichelRottleuthner) [inline sys/fatfs_diskio/Makefile]: done

Comment (MichelRottleuthner) [inline sys/fatfs_diskio/Makefile]: ping @miri64

File: sys/Makefile [hunk 36]
```diff
@@ -82,23 +82,18 @@ endif
 ifneq (,$(filter tcp,$(USEMODULE)))
     DIRS += net/transport_layer/tcp
 endif
-
```
Comment (miri64) [inline sys/Makefile]: Unrelated?

File: sys/fatfs_diskio/Makefile (hunk 35 above)
Comment (miri64) [inline sys/fatfs_diskio/Makefile]: Pong.

**Comment by miri64** [comment]
Press merge at will.

File: sys/Makefile (hunk 36 above)
Comment (MichelRottleuthner) [inline sys/Makefile]: yep - seems like I couldn't resist removing the asymmetry in this file when moving the diskio stuff^^

Comment (miri64) [inline sys/Makefile]: Are you removing the patch or shall I merge?

**Comment by BytesGalore** [comment]
beside the unrelated newline strips, looks good to me -> ACK'n'go
